### ISSUE TITLE
RTRRL 改进：以单位化资格迹导出更新（unit-trace rule），可选 Intentional TD 自适应步长

### ISSUE THREAD

**Issue by le2333** [le2333/RTRRL#37]
## 摘要

新增一条共享的 `UpdateRule`：更新量的**方向**只由资格迹方向 `ẑ` 和 TD 误差符号决定，**大小**为 `α|δ|`：

```
Δθ = α · δ · ẑ ,   ẑ = z / ‖z‖ ,   因此 ‖Δθ‖ = α·|δ|
```

`δ`、`z` 的累积、RNN 骨架梯度的合成（`upstream(actor_upward + critic_upward)`）全部保持原始向量不变，**单位化只发生在导出更新的那一步**。`α` 可为定值，也可切换为 arXiv:2604.19033 *Intentional Updates for Streaming Reinforcement Learning* 中 Intentional TD 的自适应步长。

按模块实现，不复制算法文件：新规则落在 `memo/memorax/rl/updates.py`（已有的 `UpdateRule` 共享组件层），RTRRL 只放宽它的优化器分支表。`kind: adam` 路径逐位不变。

## 动机：这不是临时起意的归一化

Intentional TD（论文 Algorithm 1）的更新是

```
ν_t  = ν_{t-1} + (1-β_ν)/(1-β_ν^t) · (g_t² - ν_{t-1})
ρ_t  = (√ν_t + ε)^{-1}
σ_t  = ⟨ρ_t g_t, g_t⟩
σ̄_t  = σ̄_{t-1} + (1-λγ)/(1-(λγ)^t) · (σ_t - σ̄_{t-1})
z_t  = λγ z_{t-1} + g_t
α_t  = η / √(σ̄_t ⟨ρ_t z_t, z_t⟩)
w_{t+1} = w_t + α_t δ̃_t ρ_t z_t
```

取 `ρ = I`（不做对角预条件）时：

```
α_t z_t = η z_t / (√σ̄_t · ‖z_t‖) = (η / √σ̄_t) · ẑ_t
```

也就是说 **Intentional TD 在 ρ=I 时本来就恰好是 `α|δ|ẑ` 的形式**，其中 `α_t = η/√σ̄_t`，`σ̄_t` 是 `‖g_t‖²` 的 `(λγ)`-EMA（带偏差校正）。

所以本 issue 提的规则不是随手加的一个归一化，而是：**把 Intentional TD 的对角预条件置为恒等后的形式**；打开自适应开关就恢复论文的步长，关掉就是固定 `η`。这一点决定了下面所有默认值的取法。

`ρ ≠ I` 时等式不再成立，方向变成 `ρz` 而非 `z`。若要逐位复现论文，导出方向取 `ρz/‖ρz‖`、步长取 `α_export = η‖ρz‖/√(σ̄⟨ρz,z⟩)`，两者之积与论文一致。故把对角缩放做成独立开关（见 §配置面）。

## 现状：改动落在哪一层

`Core._take_updates`（`memo/memorax/algorithms/rtrrl_aaai.py:785`）把三个 block 按两个 rule group 打包，交给 `rule.apply(traces, direct, state, delta=, step=, params=)`。当前 RTRRL 的两个 group 都走 `make_optax_rule`（`memo/memorax/rl/updates.py:83`）：先 `δ·z + direct`，再对 env 轴求均值，再过 Adam（torso 另有 `clip_by_global_norm`）。

`UpdateRule` 契约（`memo/memorax/rl/updates.py:49`）本来就是"只决定沿给定方向走多远"的替换点，`make_bounded_rule`（ObGD）已经是同一层的第二个实现。**所以这是模块化实现，不是新算法副本。**

不需要改动的部分（确认过）：

- `_per_stream` 里 `torso_traced = upstream(actor_upward + critic_upward)[0]` —— 原始向量合成，不动。
- `_advance_trace`（`rtrrl_aaai.py:406`）—— 迹的累积、`emphasis`、`reset_before` 不动。
- `_td_error`、normalization、`torso.follow`、differentiation —— 不动。

### 1. 归一化的作用域

`rule.apply` 收到的 `traced` 是 `{block_name: pytree}`（一个 group 内可能有多个 block；`heads` group 同时含 actor 和 critic）。三种选择：

| 作用域 | 结果 |
|---|---|
| 按 group 整体归一 | actor 与 critic 共用一个单位球，迹大的那个几乎吃掉全部预算 —— 否决 |
| **按 block 归一（推荐）** | torso / actor / critic 各得一个独立的 `α|δ|` 预算，保留各自贡献的身份 |
| 按 leaf 归一 | 退化成 sign-SGD，摧毁 block 内方向 —— 否决 |

推荐**按 block**，即对 `traced` 顶层 dict 的每一项各算一个范数。保留 `scope: block | group` 开关便于消融。

### 2. 每流（per-stream）

迹的形状是 `(num_envs, *param.shape)`。范数按流计算（沿 `axis 1..ndim` 求和），得到每流一个 `α`，最后仍按现状对 env 轴取均值。注意：多个单位向量的均值不再是单位向量 —— 这是**期望行为**（流之间不一致 ⇒ 步长自动变小）；参考设定 `num_envs=1` 下等式精确成立。

### 3. 零迹保护

`t=0` 与 reset 后梯度为零时 `‖z‖=0`。用 `z / max(‖z‖, ε)`（**不是** `z/(‖z‖+ε)`），使零迹给出精确的零更新而非爆炸值。`ε = 1e-8`。

### 4. `direct`（熵项）怎么处理

熵梯度不进迹，RTRRL 是当步直接加。三种做法：

1. **不参与归一化，直接加：`Δθ = α(δ_G·ẑ + d)`（推荐）** —— 保持 `entropy_rate` 现有语义。
2. 单独单位化后加。
3. 加进 `z` 再一起归一 —— 论文对策略是把熵折进 `g`（`g = ∇[log π + ξ·entropy·sign(Ã)]`），但那是在进迹之前，与 RTRRL 的"当步直接项"设计冲突。

选 1。**待审核确认。**

### 5. `η_f` 与 `grad_clip`

- `η_f`：torso 组的 `δ` 乘 `eta_f`，在新规则下等价于 torso 步长的一个纯倍数，语义更干净，保留。
- `torso.grad_clip`：迹部分的范数已经精确等于 `α|δ|`，全局范数裁剪基本失效（只对 `direct` 项还有作用）。建议该分支下默认 `grad_clip: 0.0`，但不移除参数。**待审核确认。**

### 6. 自适应 α 需要 `g_t`，而当前契约拿不到

论文的 `ν`、`σ` 都定义在**当步梯度** `g_t` 上。但 `_take_updates` 传给 rule 的 `traced` 是**上一步累积完的迹**（RTRRL 是"先用迹更新、再推进迹"的后置递推），当步梯度 `traced` 变量另有其名，rule 看不到。

需要把 `UpdateRule.apply` 扩一个可选关键字 `gradient=None`；Adam 与 ObGD 忽略它。这是本 issue 唯一的契约改动，影响面是 `updates.py` 两个现有 rule 的签名 + `stream_ac.py` 的调用点。

顺带记录一个与论文的既有错位（**本 issue 不修**，属"其他部分保持不变"）：RTRRL 用 `z_{t-1}` 更新，论文用 `z_t = λγz_{t-1} + g_t`。

### 7. `σ̄` 的 EMA 速率

论文里 `σ̄` 的速率就是迹的衰减 `λγ`。RTRRL 有三个不同的 `λ`（`lambda_pi` / `lambda_v` / `lambda_rnn`），而一个 group 内可能混着两个 block。所以 `σ̄` 必须**按 block** 维护，并使用该 block 自己的 `γλ`。做法：`make_unit_trace_rule` 构造时由 RTRRL 传入 `{block_name: decay}`（RTRRL 本来就知道这三个值）。这是诚实的耦合 —— `σ̄` 的定义本就依附于迹的衰减。**待审核确认**（备选：把速率做成独立超参）。

### 8. 明确不做的部分，以及一个必须记录的风险

论文另有三件与本改动相邻但独立的东西，按"其他部分保持不变"**排除在本 issue 之外**：δ 的裁剪（eq. 17）、策略侧用 `|A|` 的滑动均值做优势归一（`β_norm=0.9998`）、熵项乘 `sign(Ã)`。

**风险（需要在验收时看）**：当前 torso/heads 的步长里，Adam 的二阶矩事实上吸收了 `|δ|` 的大部分尺度波动。换成本规则后，步长与 `|δ|` **精确成正比**且没有任何上界 —— 这是唯一一处失去 `|δ|` 归一化的地方。如果短跑里看到步长尖峰，最先要补的就是 δ 裁剪或 `|A|` 归一（作为可选伴随组件），而不是回退整条规则。

## 配置面

两个正交旋钮，都默认到最简的那一档：

```yaml
space:
  torso:
    optimizer:
      kind: [unit_trace]
      unit_trace:
        scope:   [block]          # block | group
        scaling: [none]           # none | rmsprop   (ρ)
        alpha:
          kind: [fixed]           # fixed | intentional
          fixed:
            eta: [0.01]
          intentional:
            eta:    [0.1]         # 每步意图达到的 TD 误差缩减比例
            beta_nu: [0.999]
            eps:     [0.00000001]
```

- `scaling: none` + `alpha: intentional` ⇒ `α_t = η/√σ̄_t`，`ẑ = z/‖z‖` ⇒ **与论文 ρ=I 时逐位一致**，也正是本 issue 要的 `α|δ|ẑ`。
- `scaling: rmsprop` + `alpha: intentional` ⇒ `ẑ = ρz/‖ρz‖`，`α = η‖ρz‖/√(σ̄⟨ρz,z⟩)` ⇒ 与论文完整 Algorithm 1 逐位一致。
- `alpha: fixed` ⇒ `Δθ = η·δ·ẑ`。

另建议加 `magnitude: [td_error]`（备选 `sign`）：`td_error` 给 `Δθ = α·δ·ẑ`（‖Δθ‖ = α|δ|，即本文取法）；`sign` 给 `Δθ = α·sign(δ)·ẑ`（每步定长）。加这个开关是因为需求原文"仅使用 td 误差的符号"与式子 `α|δ|ẑ` 两种读法都成立，一个开关就不必二选一。**待审核确认默认值。**

## 落点

- `memo/memorax/rl/updates.py`：新增 `UnitTrace` / `FixedAlpha` / `IntentionalAlpha` 参数 dataclass、`ALPHA_FAMILY`、`make_unit_trace_rule(...)`；`UpdateRule.apply` 增加可选 `gradient=`。
- `memo/memorax/rl/__init__.py`：导出。
- `memo/memorax/algorithms/rtrrl_aaai.py`：`RTRRL_OPTIMIZERS` 放宽到 `{adam, unit_trace}`；`make_rules` 按分支分派；`_take_updates` 把当步梯度传下去。三个 block 的 `decay` 一并传入。
- `memo/memorax/algorithms/stream_ac.py`：仅跟随签名改动（新规则对它同样可用，是白拿的）。
- 新实验配置一份，照 `experiments/rtrrl issue27 fixed sample.yaml` 的 masked Hopper 形状。

## 验收

按 AGENTS.md「当两份实现必须一致时」，全部写成数值测试一起推，不靠读 diff：

1. **单元**：给定手搓的迹树，逐 block 逐流断言 `‖Δθ‖ == α|δ|`，且 `cos(Δθ, z) == 1`；零迹给出精确零更新。
2. **单元**：`scaling=none` + `alpha=intentional`，对着 Algorithm 1 的一份小 NumPy 转写逐步比对 `ν, ρ, σ, σ̄, z, α, Δw`。
3. **单元**：`scaling=rmsprop` 同上，对完整 Algorithm 1。
4. **回归**：`memo/tests/test_rtrrl_parity.py` 在 `kind: adam` 下必须全绿 —— 默认路径零行为变化。
5. **集成**：masked Hopper 短跑，读 `train/.../step_size`（`torso_step.step_size` / `heads_step.step_size` 这两个 reading 已经存在，新规则把 `α` 报到那里）与 `trace_norm`，确认步长没有尖峰。
6. 本地 `ruff check` + `pytest`，合并门仍是远端绿。

## 待审核的决定

1. §4 熵项：不参与归一化、直接加 —— 确认？
2. §5 `grad_clip` 在该分支下默认 0 —— 确认？
3. §7 `σ̄` 速率由 RTRRL 传入各 block 的 `γλ` —— 确认？（备选：独立超参）
4. §配置面 `magnitude` 默认取 `td_error`（`α|δ|ẑ`）还是 `sign`（`α·sign(δ)·ẑ`）？
5. §1 默认 `scope: block` —— 确认？

**Comment by le2333** [comment]
## Binding amendment for the formal E4 experiment

This comment supersedes the issue body's statement that phase two / IU is “not counted.” The formal experiment requires both D-RTRRL and IU-RTRRL. They may land as separate PRs, but this issue is not complete until both are implemented and independently selectable.

### D-RTRRL contract

For each parameter group `k`:

```text
Delta theta[k,t] = a[k] * sgn(delta[t]) * z[k,t] / (||z[k,t]|| + eps)
```

- The eligibility recursion is unchanged.
- `|delta|` must not set step magnitude.
- Actor/critic source-relative magnitude is preserved before synthesis.
- For shared recurrent parameters, compute the original combined recurrent eligibility/direction first, then remove only the norm of that final resultant. Never normalize actor and critic recurrent sources separately and add them afterward.
- The implementation must state and test whether the denominator is the preregistered `||z|| + eps` or the issue body's `max(||z||, eps)`. They are not numerically identical near zero; do not silently substitute one for the other. Zero trace must remain finite and produce zero update.
- Direct terms such as entropy must be specified separately. They must not accidentally make a claimed fixed-step property false; report the traced and direct contributions independently in tests.

The existing stage-one unit-trace design is the intended implementation vehicle if it satisfies every condition above.

### IU-RTRRL contract

Implement the Intentional Updates eligibility variant, including the parts required for its stability rather than copying only its final scalar formula:

- explicit separation of update direction from intended functional step size;
- the IU state/recursions for gradient scale and intended step size (`nu`, `rho`, `sigma_bar`, `alpha`) with declared initialization and bias correction;
- the paper's TD clipping/stabilization required by the IU rule;
- per-block state and the appropriate `gamma * lambda_k` rate for actor, critic and recurrent groups;
- access to the current source gradient `g_t` where IU's state update requires it, even though legacy RTRRL applies the previous accumulated trace before advancing it.

For shared recurrent parameters:

1. preserve the actor/value source-relative magnitudes;
2. form one common recurrent direction from their sum;
3. constrain the allowed shared functional step using both actor and value sensitivity;
4. never unit-normalize or independently step the two sources before summing.

Tests must include asymmetric actor/value sensitivities so an implementation that accidentally normalizes the sources separately fails clearly.

### Configuration and experiment acceptance

- `D-RTRRL` and `IU-RTRRL` are separate catalog-declared choices.
- Adam/Diag-RTRRL remains numerically unchanged.
- D-RTRRL HPO exposes the group step scales (prefer one global multiplier over frozen original ratios).
- IU-RTRRL exposes only parameters required by its equations; defaults and ranges are documented.
- TD-out (`eta * delta * normalized_or_clipped_trace`) is a named unsafe negative-control arm, capped at 10k steps or immediate non-finite termination, with the failure recorded as a result rather than hidden as infrastructure failure.
- Finite zero-norm behavior, shared-source ordering, step-magnitude invariance for D, and IU sensitivity limiting are numerical tests.

E3's N/D/G/B eligibility-readout ablation is deliberately separate in #41: E3-D retains instantaneous source magnitude, while D-RTRRL here assigns magnitude through the final update rule.

**Comment by le2333** [comment]
Implementation reuse note: PR #44 already provides the D-RTRRL/unit-trace core needed by R3.2 inside the existing memo RTRRL entry and container. Under the updated minimal plan, review it as the fixed-step saturated-limit control u_fixed = C * sign(delta) * z / (norm(z)+eps). Do not create a second implementation or container. Remaining work under this issue is to bind eta/C to the original clip threshold, add the formal 5-seed x 1M configuration/metrics, stage TD-out at 100k then 1M if safe, and integrate the identical-state R3.4 fork from #45. IU remains optional O6.

**Comment by le2333** [comment]
Acceptance follow-up after the D-RTRRL merge:

The implementation and CI coverage are present, but PR #44 did not dispatch either arm to AWS Batch. This issue therefore still needs actual short-run acceptance rather than manual inspection alone.

Use the freshly published `main@6d3c13b` CPU image:
`007122174918.dkr.ecr.eu-north-1.amazonaws.com/rtrrl@sha256:20534ef186bbc2d83fd91b0400c95c7b5529d517e071d04807b19028c3371aaf`
Build evidence: https://github.com/le2333/RTRRL/actions/runs/32150857787

Required pre-experiment acceptance:
- one short D-RTRRL run reaches completion and emits the contract metrics;
- one short TD-out control run, with any divergence/non-finite behavior recorded explicitly rather than silently treated as infrastructure failure.

I will append Batch job IDs and observed results here after the AWS session is reauthenticated.

**Comment by le2333** [comment]
Remote Batch acceptance result on main@6d3c13b / image sha256:20534ef186bbc2d83fd91b0400c95c7b5529d517e071d04807b19028c3371aaf.\n\nTD-out control:\n- job 48507f76-cece-41df-aab3-c3d8f9e61b88;\n- completed and was scored as the preregistered worst value -1e30, so the non-finite path is recorded rather than treated as infrastructure failure.\n\nD-RTRRL fixed-step arm, using an explicit collision-safe launch ID:\n- job 145f6401-14a1-4fbb-a95e-0bc48e628221;\n- worker completed and uploaded artifacts, but numerical acceptance failed;\n- with the shipped smoke value C=1.0, recurrent gradient/trace norms grew to Infinity around step 549, followed by NaN training/evaluation values; final score was -1e30.\n\nThis contradicts the smoke configuration acceptance expectation that the fixed-step arm remains finite while TD-out is the unsafe control. R3.2 formal runs remain blocked. Before formal launch, determine whether C=1.0 is an invalid placeholder or units mismatch, whether the outer clip composes incorrectly with D-RTRRL, or whether the implementation does not enforce the intended bounded realized step. Add an acceptance configuration with a justified R2-derived C and require finite realized update norms/evaluation over the safety horizon.\n\nAn earlier D-RTRRL-named job was invalidated by the concurrent launch collision now tracked in #53; the explicit-launch-ID job above is the clean result.

**Comment by le2333** [comment]
Follow-up: the fixed-step component was rerun with c=0.01, treating c as the effective learning rate of this direct non-Optax SGD-style update.

Clean acceptance run:
- job b8b849ac-91c2-413c-a79a-6e74428fc66a
- explicit launch ID acceptance-issue37-d-rtrrl-c001-20260818a
- 5k steps, seed 0
- final score 2.097857713699341
- 440 metric records, zero NaN/Infinity tokens
- finite eval return_per_step at every checkpoint: 1.444012, 2.213670, 2.110430, 2.083894, 2.097858
- recorded torso and heads step_size both 0.009999999776482582

Conclusion: D-RTRRL execution and numerical smoke acceptance pass at c=0.01. The prior c=1.0 failure reflects an excessively large effective learning rate, not an inability of the implementation to execute. Formal R3.2 should still inherit/freeze the R2-selected learning-rate scale rather than treating c as the old clip threshold by definition.
